pkg/query-service/app/clickhouseReader/reader.go

```diff
@@ -2753,3 +2762,289 @@ func (r *ClickHouseReader) GetSamplesInfoInLastHeartBeatInterval(ctx context.Con
+	// response will contain top level fields from the otel log model
+	response := model.GetFieldsResponse{
+		Selected:    constants.StaticSelectedLogFields,
+		Interesting: []model.LogField{},
+	}
+
+	// get attribute keys
+	attributes := &[]model.LogField{}
+	query := fmt.Sprintf("SELECT DISTINCT name, datatype from %s.%s group by name, datatype", r.logsDB, r.logsAttributeKeys)
+	err := r.db.Select(ctx, attributes, query)
+	if err != nil {
+		return nil, &model.ApiError{Err: err, Typ: model.ErrorInternal}
+	}
+
+	// get resource keys
+	resources := &[]model.LogField{}
+	query = fmt.Sprintf("SELECT DISTINCT name, datatype from %s.%s group by name, datatype", r.logsDB, r.logsResourceKeys)
+	err = r.db.Select(ctx, resources, query)
+	if err != nil {
+		return nil, &model.ApiError{Err: err, Typ: model.ErrorInternal}
+	}
+
+	statements := []model.CreateTableStatement{}
+	query = fmt.Sprintf("SHOW CREATE TABLE %s.%s", r.logsDB, r.logsTable)
+	err = r.db.Select(ctx, &statements, query)
+	if err != nil {
+		return nil, &model.ApiError{Err: err, Typ: model.ErrorInternal}
+	}
+
+	extractSelectedAndInterestingFields(statements[0].Statement, constants.Attributes, attributes, &response)
+	extractSelectedAndInterestingFields(statements[0].Statement, constants.Resources, resources, &response)
+	extractSelectedAndInterestingFields(statements[0].Statement, constants.Static, &constants.StaticInterestingLogFields, &response)
+
+	return &response, nil
+}
+
+func extractSelectedAndInterestingFields(tableStatement string, fieldType string, fields *[]model.LogField, response *model.GetFieldsResponse) {
+	for _, field := range *fields {
+		field.Type = fieldType
+		if strings.Contains(tableStatement, fmt.Sprintf("INDEX %s_idx", field.Name)) {
+			response.Selected = append(response.Selected, field)
+		} else {
+			response.Interesting = append(response.Interesting, field)
+		}
+	}
+}
+
+func (r *ClickHouseReader) UpdateLogField(ctx context.Context, field *model.UpdateField) *model.ApiError {
+	// if a field is selected it means that the field is indexed
+	if field.Selected {
+		// if the type is attribute or resource, create the materialized column first
+		if field.Type == constants.Attributes || field.Type == constants.Resources {
+			// create materialized
+			query := fmt.Sprintf("ALTER TABLE %s.%s ADD COLUMN IF NOT EXISTS %s %s MATERIALIZED %s_%s_value[indexOf(%s_%s_key, '%s')]", r.logsDB, r.logsTable, field.Name, field.DataType, field.Type, strings.ToLower(field.DataType), field.Type, strings.ToLower(field.DataType), field.Name)
+			err := r.db.Exec(ctx, query)
+			if err != nil {
+				return &model.ApiError{Err: err, Typ: model.ErrorInternal}
+			}
+		}
+
+		// create the index
+		if field.IndexType == nil {
+			iType := constants.DefaultLogSkipIndexType
+			field.IndexType = &iType
+		}
+		if field.IndexGranularity == nil {
+			granularity := constants.DefaultLogSkipIndexGranularity
+			field.IndexGranularity = &granularity
+		}
```

Comment:
```suggestion
		if field.IndexType == nil {
			field.IndexType = constants.DefaultLogSkipIndexType
		}
		if field.IndexGranularity == nil {
			field.IndexGranularity = constants.DefaultLogSkipIndexGranularity
		}
```

Does pointer ref/def for constants has any advantage over just using the constant values?

Reply:
So IndexType and IndexGranularity are optional fields that's why it's a pointer and you cannot assign constant values to pointer variable directly `invalid operation: cannot take address of constants.DefaultLogSkipIndexGranularity (untyped int constant 64)` .

Do you know any other methods to do this?

Reply:
I would define an enum where the first value in `UNSPECIFIED` and the following are valid values. If the value doesn't exists in the request payload it will be set to 0 = UNSPECIFIED and your conditional branching can use that if the client send value or not. This is a pattern I have seen people use and find it easier.

Reply:
Interesting will check it out.

Reply:
Have removed pointers from primitive data types as the first step.

Reply:
After changing it to value from the pointer. Enum is not required as the validator takes care of it. https://github.com/SigNoz/signoz/blob/2f178983901f4700976460ba5c9af13427ffc309/pkg/query-service/app/logs/validator.go#L30